src/main/java/baritone/pathing/path/PathExecutor.java

```diff
@@ -380,6 +380,21 @@ private boolean shouldSprintNextTick() {
         // however, descend and ascend don't request sprinting, because they don't know the context of what movement comes after it
         if (current instanceof MovementDescend) {
 
+            if (pathPosition < path.length() - 2) {
+                // keep this out of onTick, even if that means a tick of delay before it has an effect
+                IMovement next = path.movements().get(pathPosition + 1);
+                if (MovementHelper.canUseFrostWalker(ctx, next.getDest().down())) {
+                    // frostwalker only works if you cross the edge of the block on ground so in some cases we may not overshoot
+                    // Since MovementDescend can't know the next movement we have to tell it
+                    if (next instanceof MovementTraverse || next instanceof MovementParkour) {
+                        boolean couldPlaceInstead = Baritone.settings().allowPlace.value && behavior.baritone.getInventoryBehavior().hasGenericThrowaway() && next instanceof MovementParkour; // traverse doesn't react fast enough
+                        boolean sameFlatDirection = current.getDirection().up().crossProduct(next.getDirection()).equals(BlockPos.ORIGIN); // here's why you learn maths in school
```

Comment:
so what is this cross product actually doing? `a.crossProduct(b).equals(ORIGIN)` does what exactly? checks if a and b are the same? checks if they are in the same plane but of different magnitude?

Reply:
It's there to check whether the two movements are in the same horizontal direction and have a height difference of one (because descend always goes down and we are not interested in parkour ascend, only horizontal would be `.dot(UP) == 0` instead of `.equals(ORIGIN)`).
Theoretically there's also `a.crossProduct(a.scale(-n)).equals(ORIGIN)`, but you can't jump in the direction you came from (because you would jump through the `src` of the descend).

Reply:
Gotcha. `current.getDirection()` might be `(+n, -1, 0)` and `next.getDirection()` might be `(1, 0, 0)`, then `sameFlatDirection` would be true because `current.getDirection().up()` would be `(+n, 0, 0)`, and the cross product of `(+n, 0, 0)` and `(1, 0, 0)` will be `(0, 0, 0)` because both of those vectors are in the exact same direction and don't form a plane, thus the cross product is all zero?

Please add a comment explaining all that 😉

Reply:
Actually traverse can go through `current.getSrc().down()` if `allowDownward` is or it has to place `current.getSrc().down(2)`. I've added a check that should make this work with whatever crazy movement the future might throw at it.

Reply:
Actually only Vec3d has the required method. Will force push a less universal check for `MovementTraverse` only.